tests/functional/stn/ajax.test.js

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2020 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+const testDriver = require('../../../tools/jil/index')
+
+let supported = testDriver.Matcher.withFeature('stn')
+
+testDriver.test('session trace resources', supported, function (t, browser, router) {
+  let assetURL = router.assetURL('stn/ajax-disabled.html', {
+    loader: 'full',
+    init: {
+      session_trace: {
+        harvestTimeSeconds: 5
+      },
+      ajax: {
+        enabled: false
+      },
+      page_view_timing: {
+        enabled: false
+      }
+    }
+  })
+
+  let loadPromise = browser.safeGet(assetURL)
+  let rumPromise = router.expectRum()
+  let resourcePromise = router.expectResources()
+
+  Promise.all([resourcePromise, loadPromise, rumPromise]).then(([result]) => {
+    t.equal(result.res.statusCode, 200, 'server responded with 200')
+
+    // trigger an XHR call after
+    var clickPromise = browser
+      .elementByCssSelector('body')
+      .click()
+
+    resourcePromise = router.expectResources()
+
+    return Promise.all([resourcePromise, clickPromise])
+  })
+  .then(([result]) => {
+    t.equal(result.res.statusCode, 200, 'server responded with 200')
+
+    const body = result.body
+    const harvestBody = JSON.parse(body).res
+    const loadNodes = harvestBody.filter(function (node) { return node.t === 'event' && node.n === 'load' || node.n === 'readystatechange' })
+    t.notOk(loadNodes.length > 0, 'XMLHttpRequest nodes not captured when ajax instrumentation is disabled')
+
+    t.end()
+  }).catch(fail)
+
+  function fail (err) {
+    t.error(err)
+    t.end()
+  }
+})
+
+testDriver.test('session trace ajax deny list', supported, function (t, browser, router) {
+  let assetURL = router.assetURL('stn/ajax-disabled.html', {
+    loader: 'full',
+    init: {
+      session_trace: {
+        harvestTimeSeconds: 5
+      },
+      ajax: {
+        harvestTimeSeconds: 2,
+        enabled: true,
+        deny_list:[router.router.assetServer.host]
+      },
+      page_view_timing: {
+        enabled: false
+      }
+    }
+  })
+
+  let loadPromise = browser.safeGet(assetURL)
+  let rumPromise = router.expectRum()
+  let resourcePromise = router.expectResources()
+  const ajaxPromise = router.expectBeaconRequest(
```

Comment:
can you use `expectXHRMetrics()` here so that we aren't crossing error payloads? (err, xhr, cm, sm all share that stream, so the above listener prevents those from polluting)

Reply:
I had to use `expectBeaconRequest` so I could set a custom timeout.

Reply:
IMO we should instead just extend all the promises to accept+pass on a custom timeout, because the issue still remains that if the agent happens to send a supportability metric before the xhrMetrics payload, it will be an inconsistent test.

Reply:
Agree. That is done in the fastify rewrite of the test server.